src/storage/table/variant_column_data.cpp

```diff
@@ -0,0 +1,618 @@
+	auto &unshredded_stats = VariantStats::GetUnshreddedStats(transformed_stats);
+	auto &shredded_stats = VariantStats::GetShreddedStats(transformed_stats);
+
+	VariantShreddedAppendInput append_data {*unshredded,           *shredded,        unshredded_append_state,
+	                                        shredded_append_state, unshredded_stats, shredded_stats};
+	idx_t vector_index = 0;
+	for (idx_t scanned = 0; scanned < total_count; scanned += STANDARD_VECTOR_SIZE) {
+		scan_chunk.Reset();
+		auto to_scan = MinValue(total_count - scanned, static_cast<idx_t>(STANDARD_VECTOR_SIZE));
+		ScanCommitted(vector_index++, scan_state, scan_vector, false, to_scan);
+		append_chunk.Reset();
+
+		AppendShredded(scan_vector, append_vector, to_scan, append_data);
+	}
+	stats->statistics = std::move(transformed_stats);
+	return ret;
+}
+
+LogicalType VariantColumnData::GetShreddedType() {
+	VariantShreddingStats variant_stats;
+
+	//! scan_chunk
+	DataChunk scan_chunk;
+	scan_chunk.Initialize(Allocator::DefaultAllocator(), {LogicalType::VARIANT()}, STANDARD_VECTOR_SIZE);
+	auto &scan_vector = scan_chunk.data[0];
+
+	ColumnScanState scan_state(nullptr);
+	InitializeScan(scan_state);
+	idx_t total_count = count.load();
+	idx_t vector_index = 0;
+	for (idx_t scanned = 0; scanned < total_count; scanned += STANDARD_VECTOR_SIZE) {
+		scan_chunk.Reset();
+		auto to_scan = MinValue(total_count - scanned, static_cast<idx_t>(STANDARD_VECTOR_SIZE));
+		ScanCommitted(vector_index++, scan_state, scan_vector, false, to_scan);
+		variant_stats.Update(scan_vector, to_scan);
+	}
+
+	return variant_stats.GetShreddedType();
+}
+
+static bool EnableShredding(int64_t minimum_size, idx_t current_size) {
+	if (minimum_size == -1) {
+		//! Shredding is entirely disabled
+		return false;
+	}
+	return current_size >= static_cast<idx_t>(minimum_size);
+}
+
+unique_ptr<ColumnCheckpointState> VariantColumnData::Checkpoint(const RowGroup &row_group,
+                                                                ColumnCheckpointInfo &checkpoint_info) {
+	auto &partial_block_manager = checkpoint_info.GetPartialBlockManager();
+	auto checkpoint_state = make_uniq<VariantColumnCheckpointState>(row_group, *this, partial_block_manager);
+	checkpoint_state->validity_state = validity->Checkpoint(row_group, checkpoint_info);
+
+	auto &table_info = row_group.GetTableInfo();
+	auto &db = table_info.GetDB();
+	auto &config_options = DBConfig::Get(db).options;
+
+	if (!HasAnyChanges() || !EnableShredding(config_options.variant_minimum_shredding_size, row_group.count.load())) {
+		for (idx_t i = 0; i < sub_columns.size(); i++) {
+			checkpoint_state->child_states.push_back(sub_columns[i]->Checkpoint(row_group, checkpoint_info));
+		}
+		return std::move(checkpoint_state);
+	}
+
+	LogicalType shredded_type;
+	if (config_options.force_variant_shredding.id() != LogicalTypeId::INVALID) {
+		shredded_type = config_options.force_variant_shredding;
+	} else {
+		shredded_type = GetShreddedType();
```

Comment:
Potential optimization: don't rescan and reshred everything, keep the current shredding (if IsShredded()), shredded type doesn't change, can directly reuse the 'shredded' sub_column (something along those lines)